### ISSUE TITLE
[MINSTALL-116] Attached artifacts installed with default name, only one is actually installed

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-install-plugin#210]
**[Boaz Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=boazlev)** opened **[MINSTALL-116](https://issues.apache.org/jira/browse/MINSTALL-116?redirect=false)** and commented

InstallMojo.execute() calls (for every attached artifact) to
installer.install( attached.getFile(), attached, localRepository );
The source file name is based on finalName, but destination file name is based on the localRepository.pathOf(File) method, which in the org.apache.maven.artifact.repository.layout.DefaultRepositoryLayout implementation returns a result based on artifactId, version etc.
In short, this means that if the artifactId is "X" and the version is "0.0.1-SNAPSHOT" then an artifact with a name "a.zip" will be installed as "X-0.0.1-SNAPSHOT.zip"

Since all attached artifacts are installed with the default name, in case there are more than on attached artifacts - only the last is actually installed, all others are getting lost in the process.

I would expect that any attached artifact would be installed using it's finalName, exactly as it is on the target folder.

---

**Affects:** 2.5.2

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

When installing (or deploying) an artifact, is must have a predictable path, otherwise Maven can't download it in the future. Based on the groupId, artifactId, version and type (which is translated to a file-extension) and optional classifier Maven can build that path.

How would a `<dependency>` look like if a custom final name was used?

> Since all attached artifacts are installed with the default name, in case there are more than on attached artifacts - only the last is actually installed, all others are getting lost in the process.

Attached artifacts **must** use a classifier to recognize the different attachments.

**Comment by jira-importer** [comment]
**[Boaz Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=boazlev)** commented

I agree that in case an artifact should be downloaded in the future, then it must have a predictable name, as described. But in fact, once I use the assembly plugin with more than one assembly instruction, in order to group couple of resources in separated zip files, for example - those are not tend to be downloaded rather be used as part of the release (e.g. be unzipped at different production machines).
Currently, only one zip file is uploaded! How one should handle this situation, other than writing my own version of maven-install-plugin, that does install ALL zip files in their finalName?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Then there's something wrong with either your maven-assembly-plugin configuration or your descriptors. Be aware that the id of the descriptor as defined in the XML is used as the classifier, see http://maven.apache.org/plugins/maven-assembly-plugin/assembly.html#class_assembly

**Comment by jira-importer** [comment]
**[Boaz Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=boazlev)** commented

The maven-assembly-plugin configuration and the descriptors are O.K. their \<id> is properly configured, yet ignored as configured in pom.xml, and the expected artifacts are generated in the target: a.zip and b.zip.What was not expected? ONLY b.zip was installed, and with different name...
